massa-execution-worker/src/execution.rs

```diff
@@ -1260,15 +1381,49 @@ impl ExecutionState {
             self.mip_store.clone(),
         );
 
+        // Deferred calls
+        let calls = execution_context.deferred_calls_advance_slot(
+            slot.clone(),
+            DEFERRED_CALL_MAX_FUTURE_SLOTS,
+            self.config.thread_count,
+        );
+
         // Get asynchronous messages to execute
         let messages = execution_context.take_async_batch(
-            self.config.max_async_gas,
+            self.config.max_async_gas.saturating_sub(calls.slot_gas),
             self.config.async_msg_cst_gas_cost,
         );
 
         // Apply the created execution context for slot execution
         *context_guard!(self) = execution_context;
 
+        for (id, call) in calls.slot_calls {
+            if call.cancelled {
```

Comment:
this check can be removed if `advance_slot` in `speculative_deferred_call.rs` filters cancelled calls.